File: MagicalRecordAndMantleСomparison/MagicalRecordAndMantleСomparison/MasterViewController.m
//
//  MasterViewController.m
//  MagicalRecordAndMantleСomparison
//
//  Created by Vladimir Goncharov on 10.03.15.
//  Copyright (c) 2015 FlatStack. All rights reserved.
//

#import "MasterViewController.h"
#import "DetailViewController.h"

#if TEST_MAGICAL_RECORD

#import <MagicalRecord/CoreData+MagicalRecord.h>
#import "Person.h"
#import "Address.h"

#else

#import "CoreDataManager.h"
#import "Bug.h"
#import "GHBug.h"
#import "Issue.h"
#import "GHIssue.h"
#import "Question.h"
#import "GHQuestion.h"
#import "User.h"
#import "GHUser.h"

#endif

@interface MasterViewController ()

@property (nonatomic, strong) NSMutableArray *objects;

@end

@implementation MasterViewController

- (void)viewDidLoad
{
    [super viewDidLoad];
    // Do any additional setup after loading the view, typically from a nib.
    self.navigationItem.rightBarButtonItem = self.editButtonItem;
    
#if TEST_MAGICAL_RECORD
    self.objects    = [[Person MR_findAll] mutableCopy];
#else
    
    NSError *error = nil;
    NSFetchRequest *request = [[NSFetchRequest alloc] initWithEntityName:[Issue entityName]];
    
    self.objects =
    [[[CoreDataManager sharedManager].managedObjectContext executeFetchRequest:request
                                                                        error:&error] mutableCopy];
#endif
}

#pragma mark - Segues

- (void)prepareForSegue:(UIStoryboardSegue *)segue
                 sender:(id)sender
{
    if ([[segue identifier] isEqualToString:@"showDetail"])
    {
        NSIndexPath *indexPath = [self.tableView indexPathForSelectedRow];
        NSDate *object = self.objects[indexPath.row];
        [[segue destinationViewController] setDetailItem:object];
    }
}

#pragma mark - Table View

- (NSInteger)numberOfSectionsInTableView:(UITableView *)tableView
{
    return 1;
}

- (NSInteger)tableView:(UITableView *)tableView
 numberOfRowsInSection:(NSInteger)section
{
    return self.objects.count;
}

- (UITableViewCell *)tableView:(UITableView *)tableView
         cellForRowAtIndexPath:(NSIndexPath *)indexPath
{
    UITableViewCell *cell = [tableView dequeueReusableCellWithIdentifier:@"Cell" forIndexPath:indexPath];
#if TEST_MAGICAL_RECORD
    Person *object = self.objects[indexPath.row];
    cell.textLabel.text = [object fullName];
#else
    Issue *object = self.objects[indexPath.row];
    cell.textLabel.text = object.title;
#endif
    return cell;
}

- (BOOL)tableView:(UITableView *)tableView
canEditRowAtIndexPath:(NSIndexPath *)indexPath
{
    // Return NO if you do not want the specified item to be editable.
    return YES;
}

- (UITableViewCellEditingStyle)tableView:(UITableView *)tableView
           editingStyleForRowAtIndexPath:(NSIndexPath *)indexPath
{
    return UITableViewCellEditingStyleDelete;
}

- (void)tableView:(UITableView *)tableView
commitEditingStyle:(UITableViewCellEditingStyle)editingStyle
forRowAtIndexPath:(NSIndexPath *)indexPath
{
    if (editingStyle == UITableViewCellEditingStyleDelete)
    {
#if TEST_MAGICAL_RECORD
        __weak typeof(self) wself   = self;
        [MagicalRecord saveWithBlock:^(NSManagedObjectContext *localContext) {
            Person *person = [wself.objects[indexPath.row] MR_inContext:localContext];
            [person MR_deleteEntity];
        } completion:^(BOOL success, NSError *error) {
            if (success)
            {
                [wself.objects removeObjectAtIndex:indexPath.row];
                [tableView deleteRowsAtIndexPaths:@[indexPath] withRowAnimation:UITableViewRowAnimationFade];
            }
        }];
#else
        Issue *object = self.objects[indexPath.row];
        [[[CoreDataManager sharedManager] managedObjectContext] deleteObject:object];
        [[CoreDataManager sharedManager] saveContext];
        
        [self.objects removeObjectAtIndex:indexPath.row];
        [tableView deleteRowsAtIndexPaths:@[indexPath] withRowAnimation:UITableViewRowAnimationFade];
#endif
    } 
}

@end

File: RestKitExample/RestKitExample/MasterViewController.swift
//
//  MasterViewController.swift
//  RestKitExample
//
//  Created by Vladimir Goncharov on 28.09.15.
//  Copyright © 2015 FlatStack. All rights reserved.
//

import UIKit
import CoreData

class MasterViewController: UITableViewController {

    var fetchedResultsController: NSFetchedResultsController {
        if _fetchedResultsController != nil {
            return _fetchedResultsController!
        }
        
        _fetchedResultsController = Calendar.MR_fetchAllSortedBy(CalendarAttributes.name.rawValue, ascending: true, withPredicate: nil, groupBy: nil, delegate: self, inContext: NSManagedObjectContext.MR_defaultContext())!
        
        do {
            try _fetchedResultsController!.performFetch()
        } catch let error as NSError {
            // Replace this implementation with code to handle the error appropriately.
            // abort() causes the application to generate a crash log and terminate. You should not use this function in a shipping application, although it may be useful during development.
            print("Unresolved error \(error), \(error.userInfo)")
            abort()
        }
        
        return _fetchedResultsController!
    }
    var _fetchedResultsController: NSFetchedResultsController? = nil
    
    //MARK: -
    override func viewDidLoad() {
        
        super.viewDidLoad()
        
        self.tableView.estimatedRowHeight = 44.0
        self.tableView.rowHeight = UITableViewAutomaticDimension
        
        Calendar.API_getListOfCalendar(nil) { (operation, error) -> Void in
            print(error.userInfo[RKObjectMapperErrorObjectsKey]?.firstObject)
        }
        
        // Do any additional setup after loading the view, typically from a nib.
        self.navigationItem.leftBarButtonItem = self.editButtonItem()
    }

    // MARK: - Segues

    override func prepareForSegue(segue: UIStoryboardSegue, sender: AnyObject?) {
        if segue.identifier == "showDetail" {
            if let indexPath = self.tableView.indexPathForSelectedRow {
            let object = self.fetchedResultsController.objectAtIndexPath(indexPath) as? Calendar
                let controller = segue.destinationViewController as! DetailViewController
                controller.detailItem = object
                controller.navigationItem.leftItemsSupplementBackButton = true
            }
        } else {
            super.prepareForSegue(segue, sender: sender)
        }
    }
}

//MARK: -
extension MasterViewController {
    
    override func numberOfSectionsInTableView(tableView: UITableView) -> Int {
        return self.fetchedResultsController.sections?.count ?? 0
    }
    
    override func tableView(tableView: UITableView, numberOfRowsInSection section: Int) -> Int {
        let sectionInfo = self.fetchedResultsController.sections![section]
        return sectionInfo.numberOfObjects
    }
    
    override func tableView(tableView: UITableView, cellForRowAtIndexPath indexPath: NSIndexPath) -> UITableViewCell {
        let cell = tableView.dequeueReusableCellWithIdentifier("Cell", forIndexPath: indexPath)
        self.configureCell(cell, atIndexPath: indexPath)
        return cell
    }
    
    override func tableView(tableView: UITableView, canEditRowAtIndexPath indexPath: NSIndexPath) -> Bool {
        // Return false if you do not want the specified item to be editable.
        return true
    }
    
    override func tableView(tableView: UITableView, commitEditingStyle editingStyle: UITableViewCellEditingStyle, forRowAtIndexPath indexPath: NSIndexPath) {
        if editingStyle == .Delete {
            MagicalRecord.saveWithBlockAndWait({ (context) -> Void in
                let object = (self.fetchedResultsController.objectAtIndexPath(indexPath) as! NSManagedObject).MR_inContext(context)
                object.MR_deleteEntity()
            })
        }
    }
    
    func configureCell(cell: UITableViewCell, atIndexPath indexPath: NSIndexPath) {
        let object = self.fetchedResultsController.objectAtIndexPath(indexPath) as! Calendar
        cell.textLabel?.text = object.name
        cell.detailTextLabel?.text = object.detailDescription
    }
}


//MARK: - NSFetchedResultsControllerDelegate
extension MasterViewController : NSFetchedResultsControllerDelegate {
    
    func controllerWillChangeContent(controller: NSFetchedResultsController) {
        self.tableView.beginUpdates()
    }
    
    func controller(controller: NSFetchedResultsController, didChangeSection sectionInfo: NSFetchedResultsSectionInfo, atIndex sectionIndex: Int, forChangeType type: NSFetchedResultsChangeType) {
        switch type {
        case .Insert:
            self.tableView.insertSections(NSIndexSet(index: sectionIndex), withRowAnimation: .Fade)
        case .Delete:
            self.tableView.deleteSections(NSIndexSet(index: sectionIndex), withRowAnimation: .Fade)
        default:
            return
        }
    }
    
    func controller(controller: NSFetchedResultsController, didChangeObject anObject: AnyObject, atIndexPath indexPath: NSIndexPath?, forChangeType type: NSFetchedResultsChangeType, newIndexPath: NSIndexPath?) {
        switch type {
        case .Insert:
            tableView.insertRowsAtIndexPaths([newIndexPath!], withRowAnimation: .Fade)
        case .Delete:
            tableView.deleteRowsAtIndexPaths([indexPath!], withRowAnimation: .Fade)
        case .Update:
            self.configureCell(tableView.cellForRowAtIndexPath(indexPath!)!, atIndexPath: indexPath!)
        case .Move:
            tableView.deleteRowsAtIndexPaths([indexPath!], withRowAnimation: .Fade)
            tableView.insertRowsAtIndexPaths([newIndexPath!], withRowAnimation: .Fade)
        }
    }
    
    func controllerDidChangeContent(controller: NSFetchedResultsController) {
        self.tableView.endUpdates()
    }
}
